### PR TITLE
feat(sparse): supernodal LU numeric with threshold partial pivoting (#182)

### LINKED ISSUE #182: feat(sparse): supernodal LU numeric with threshold partial pivoting
Phase 2 (the core) of native supernodal LU (SuperLU) performance parity (see docs/plans/native-superlu-performance.md). Part of the native SuperLU performance epic.

## Work
- Left-looking supernode-panel numeric with **panel partial pivoting** and **relaxed supernodes**.
- Route the dense block update through the mixed-precision GEMM path (`mult<Accumulator>` / `accumulator_traits`), reusing the dense-panel machinery from #178.
- Emit a standard CSC L/U so the existing solve and iterative refinement are reused.

## Acceptance
- Factorization correct vs SuperLU on the suite (residual parity); native factor within ~3x of SuperLU on a dense-structured matrix.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

**Phase 2 — the core** of the native-SuperLU epic (#186), closes #182. Adds `include/mtl/sparse/factorization/supernodal_lu.hpp`: a left-looking Gilbert–Peierls LU (`PA = LU`) that groups columns of L into **supernodes** so the dominant elimination work is a **dense block update (TRSV + GEMV)** — the granularity where mixed precision pays off.

The hard part — pivoting vs supernodes — is handled by forming supernodes **dynamically** during the pivoted factorization (a static structure is impossible with pivoting), capped by the Phase-1 column-etree relaxed-supernode boundaries.

## How it works

Lifts the proven machinery from scalar `sparse_lu` (reach DFS, Eisenstat–Liu pruning, threshold partial pivoting, CSC L/U layout, `solve`) and upgrades only the inner update:
- **Closed supernodes** (≥1 col, finalized) are applied as one dense block via `accumulator_traits` (TRSV on the unit-lower diagonal block + GEMV on the off-diagonal — products in `Value`, accumulated in `Accumulator`).
- The **current open supernode** (a few columns) uses the proven scalar SAXPY.
- **Key correctness insight:** after the Phase-1 postorder, increasing column index is a valid topological order, so a supernode's columns are contiguous and the block update is correct. A column extends the open supernode iff its below-diagonal rows nest under the previous column's (`cand_k == cand_{k-1} \ {piv_k}`), within the relaxed boundary + a width cap.
- Emits standard CSC `L`/`U` + `row_perm`, so `solve()` and `iterative_refine` are reused. Adds `supernodal_lu_solve` and `supernodal_lu_solve_refined<FactorValue, Accumulator>`.

## Validation (external SuperLU oracle, local)

| matrix | n | max\|x_native − x_superlu\| | native nsuper | native fill | SuperLU fill |
|--------|--:|--:|--:|--:|--:|
| rand 0.05 | 100 | 1.2e-17 | 58 | 3,366 | 5,460 |
| rand 0.03 | 200 | 1.3e-17 | 92 | 18,412 | 21,870 |
| dense | 64 | 6.1e-18 | **1** | 4,160 | 4,160 |
| dense | 120 | 4.3e-18 | 2 | 14,520 | 14,520 |

Solutions match SuperLU to machine precision; dense ⇒ a single supernode (block path fully exercised); fill is competitive (≤ SuperLU; equal on dense).

## Tests

`tests/unit/sparse/test_supernodal_lu.cpp` (Catch2, 122 assertions): correctness vs scalar `sparse_lu` (random/dense/conv-diffusion), pivoting on a zero diagonal, supernode structure (diagonal ⇒ all singletons, dense ⇒ merges), mixed-precision accumulator (double-acc decisively beats float), float-factor + double-residual iterative refinement, 1×1 + singular. Full sparse suite 28/28; warning-clean under GCC + Clang `-Wall -Wextra`.

## Scope / notes

- Per the approved plan: **column-at-a-time target** with generic dense block kernels. Multi-column panels, BLAS-3-tuned kernels, and relaxed-supernode size tuning are **Phase 3 (#183)**.
- Validating against larger 2-D grids surfaced a **pre-existing** COLAMD/AMD out-of-bounds bug (filed as **#189**), independent of this kernel — `supernodal_lu` is correct on every input COLAMD orders successfully.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added supernodal LU factorization support for sparse matrices, including direct solve and mixed-precision refined solve options.
  * Expanded the main sparse include to expose the new factorization tools through a single header.

* **Bug Fixes**
  * Added validation for matrix size and permutation inputs.
  * Zero-pivot and singular cases now fail with a clear runtime error instead of producing invalid results.

* **Tests**
  * Added unit coverage for correctness, pivoting, supernode behavior, mixed-precision accuracy, iterative refinement, and edge cases.

### CODE REVIEW

File: include/mtl/sparse/factorization/supernodal_lu.hpp [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+/// \param threshold pivoting threshold in (0,1]; 1 = full partial pivoting.
+/// \param max_super maximum supernode width (relaxation cap).
+/// \throws std::runtime_error on a singular (zero) pivot.
+template <typename Value, typename Parameters, typename Accumulator = Value>
+    requires OrderedField<Value>
+supernodal_lu_factor<Value> supernodal_lu_numeric(
+    const mat::compressed2D<Value, Parameters>& A,
+    const supernodal_lu_symbolic& sym,
+    Value threshold = Value{1},
+    std::size_t max_super = 64)
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_lu.hpp:142]: Fixed — added an up-front `threshold ∈ (0,1]` guard before any elimination.

File: include/mtl/sparse/factorization/supernodal_lu.hpp [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+{
+    using size_type = std::size_t;
+    using std::abs;
+    using AT = accumulator_traits<Accumulator, Value>;
+    const size_type n = sym.n;
+    if (A.num_rows() != n || A.num_cols() != n)
+        throw std::invalid_argument("supernodal_lu_numeric: dimensions do not match symbolic analysis");
+
+    auto C = util::crs_to_csc(util::column_permute(A, sym.col_perm));
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_lu.hpp:160]: Fixed — validate the symbolic object (col_perm a valid size-n permutation; col_pinv/col_super size n) before indexing, mirroring sparse_lu_numeric.

File: include/mtl/sparse/factorization/supernodal_lu.hpp [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    using idx32 = std::int32_t;
+    std::vector<idx32>          pinv(n, -1);
+    std::vector<Accumulator>    x(n);
+    std::vector<idx32>          xi(n);
+    std::vector<std::ptrdiff_t> pstack(n);
+    std::vector<char>           marked(n, 0);
+    std::vector<idx32>          reach_nodes; reach_nodes.reserve(n);
+    std::vector<size_type>      Lp(n + 1, 0), Up(n + 1, 0), xprune(n, 0);
+    std::vector<idx32> Li; std::vector<Value> Lx;
+    std::vector<idx32> Ui; std::vector<Value> Ux;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_lu.hpp:171]: The int32 row-index storage is the documented memory-bandwidth choice inherited from sparse_lu (all stored values are row/node ids in [0,n) with a -1 sentinel). Rather than widen (and lose that), I added a fail-safe guard that throws when n exceeds the int32 range, so oversized inputs fail cleanly instead of overflowing.

File: include/mtl/sparse/factorization/supernodal_lu.hpp [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+    // Supernode bookkeeping.
+    std::vector<detail::lu_snode<Value>> snodes;   // closed supernodes
+    std::vector<idx32> col_csuper(n, -1);          // col -> closed supernode id (or -1 if open/unset)
+    std::vector<size_type> lsuper_first;           // boundaries, recorded as supernodes close
+    size_type open_sf = 0;                         // first column of the currently-open supernode
+    std::vector<size_type> prev_cand;              // sorted orig candidate rows of column k-1
+
+    std::vector<idx32> contrib_super;              // closed supernodes hitting column k (dedup)
+    std::vector<char>  super_seen;                 // marker per closed supernode
+    std::vector<Value> seg;                        // dense supernode segment workspace
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_lu.hpp]: Good catch on the intent. The literal change (seg as Accumulator feeding add_product) wouldn't hold generically — add_product takes Value operands (so a quire Acc can't be the multiplier). Fixed it correctly: the segment is now an Accumulator accumulated via add_product through the TRSV, and each pivot is rounded to Value only at the point it's used as a multiplier. So the intra-supernode solve now stays in accumulator precision.

File: include/mtl/sparse/factorization/supernodal_lu.hpp [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+                if (ci != cand.size()) join = false;     // cur_pivot wasn't in prev_cand
+            }
+            if (!join) { close_supernode(k, prev_cand); open_sf = k; }
+        }
+        prev_cand = std::move(cand);
+    }
+    Lp[n] = Li.size();
+    Up[n] = Ui.size();
+    close_supernode(n, prev_cand);                 // close the final open supernode
+    lsuper_first.push_back(n);
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_lu.hpp]: Fixed — empty matrix now records zero supernodes (no [0,0] boundary); added an empty-matrix test asserting nsuper()==0.

File: include/mtl/sparse/factorization/supernodal_lu.hpp [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+    // Re-type the sparse matrix into the factor precision, preserving sparsity
+    // (mtl::convert is for dense tensors and would densify a compressed2D).
+    const std::size_t nnz = A.nnz();
+    std::vector<size_type> starts(A.ref_major().begin(), A.ref_major().end());
+    std::vector<size_type> idx(A.ref_minor().begin(), A.ref_minor().end());
+    std::vector<FactorValue> dat(nnz);
+    const auto& src = A.ref_data();
+    for (std::size_t kk = 0; kk < nnz; ++kk) dat[kk] = static_cast<FactorValue>(src[kk]);
+    mat::compressed2D<FactorValue> Af(A.num_rows(), A.num_cols(), nnz,
+                                      starts.data(), idx.data(), dat.data());
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_lu.hpp]: Fixed — Af is now compressed2D<FactorValue, Parameters>, preserving the source layout/size_type.

File: tests/unit/sparse/test_supernodal_lu.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    // Storage double (accurate solve) so the residual reflects FACTOR quality.
+    auto fac_f = factorization::supernodal_lu_numeric<double, decltype(A)::param_type, float>(A, sym);
+    auto fac_d = factorization::supernodal_lu_numeric<double, decltype(A)::param_type, double>(A, sym);
+    vec::dense_vector<double> xf(n, 0.0), xd(n, 0.0);
+    fac_f.solve(xf, b); fac_d.solve(xd, b);
+    double res_f = rel_residual(A, xf, b), res_d = rel_residual(A, xd, b);
+
+    REQUIRE(std::isfinite(res_f));
+    REQUIRE(res_d < 1e-12);
+    REQUIRE(res_d < res_f);                        // double accumulation decisively better
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply tests/unit/sparse/test_supernodal_lu.cpp]: Fixed — loosened the absolute gates (res_d < 5e-11, IR < 1e-9) while keeping the relative-improvement checks, and added an explicit clear-separation assertion (res_f > 1e3·res_d) so the test still proves the accumulator effect.
